File: ios/OpenPatent/Core/Networking/OrchestratorRepository.swift
import Foundation

protocol OrchestratorRepositoryProtocol {
    func createPlan(userRequest: String, technology: String) async throws -> OrchestrationPlan
    func executePlan(planId: String) async throws -> [OrchestrationPlan.OrchestrationTask]
    func getTemplates() async throws -> [String]
    func getThinking(sessionId: String) async throws -> [[String: String]]
}

final class OrchestratorRepository: OrchestratorRepositoryProtocol {
    private let api: ApiClientProtocol
    
    init(api: ApiClientProtocol) {
        self.api = api
    }
    
    func createPlan(userRequest: String, technology: String) async throws -> OrchestrationPlan {
        try await api.request(.createPlan, body: OrchestratorRequest(userRequest: userRequest, technology: technology, projectPath: nil))
    }
    
    func executePlan(planId: String) async throws -> [OrchestrationPlan.OrchestrationTask] {
        struct ExecuteResponse: Decodable {
            let tasks: [OrchestrationPlan.OrchestrationTask]
        }
        let response: ExecuteResponse = try await api.request(.executePlan, body: OrchestratorExecuteRequest(planId: planId))
        return response.tasks
    }
    
    func getTemplates() async throws -> [String] {
        struct TemplatesResponse: Decodable {
            let templates: [String]
        }
        let response: TemplatesResponse = try await api.request(.getTemplates, body: nil)
        return response.templates
    }
    
    func getThinking(sessionId: String) async throws -> [[String: String]] {
        struct ThinkingResponse: Decodable {
            let thinking: [[String: String]]
        }
        let response: ThinkingResponse = try await api.request(.getThinking(sessionId: sessionId), body: nil)
        return response.thinking
    }
}

File: android/app/src/main/java/ai/openpatent/data/repository/OrchestratorRepository.kt
package ai.openpatent.data.repository

import ai.openpatent.data.model.*
import ai.openpatent.data.remote.api.OrchestratorApi
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class OrchestratorRepository @Inject constructor(
    private val orchestratorApi: OrchestratorApi
) {
    suspend fun createPlan(
        request: String,
        technology: String = "software",
        sessionId: String? = null
    ): Result<OrchestratorPlan> {
        return try {
            val planRequest = OrchestratorPlanRequest(request, technology, sessionId)
            val response = orchestratorApi.createPlan(planRequest)
            if (response.success && response.plan != null) {
                Result.success(response.plan)
            } else {
                Result.failure(Exception(response.error ?: "Failed to create plan"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun executePlan(
        request: String,
        technology: String = "software",
        sessionId: String? = null,
        planId: String? = null
    ): Result<OrchestratorExecuteResponse> {
        return try {
            val execRequest = OrchestratorExecuteRequest(request, technology, sessionId, planId)
            val response = orchestratorApi.executePlan(execRequest)
            if (response.success) {
                Result.success(response)
            } else {
                Result.failure(Exception(response.error ?: "Failed to execute plan"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun getTemplates(): Result<List<OrchestratorTemplate>> {
        return try {
            val response = orchestratorApi.getTemplates()
            Result.success(response.templates)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun getThinkingLog(sessionId: String): Result<ThinkingLogResponse> {
        return try {
            val response = orchestratorApi.getThinkingLog(sessionId)
            Result.success(response)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
